django/apps/registry.py

```diff
@@ -69,53 +70,58 @@ def populate(self, installed_apps=None):
         # populate() might be called by two threads in parallel on servers
         # that create threads before initializing the WSGI callable.
         with self._lock:
-            if self.ready:
-                return
-
-            # app_config should be pristine, otherwise the code below won't
-            # guarantee that the order matches the order in INSTALLED_APPS.
-            if self.app_configs:
-                raise RuntimeError("populate() isn't reentrant")
-
-            # Phase 1: initialize app configs and import app modules.
-            for entry in installed_apps:
-                if isinstance(entry, AppConfig):
-                    app_config = entry
-                else:
-                    app_config = AppConfig.create(entry)
-                if app_config.label in self.app_configs:
+            try:
```

Comment:
I needed to reset the _populate_running flag because I misused `set_installed_apps` in the test. Using the context manager `with self.settings(INSTALLED_APPS=[...]):` calls `unset_installed_apps` and restores the proper installed apps.